### DISCUSSION TITLE
QUESTION: Keep columns when using PARTITIONED BY with SQL

### DISCUSSION THREAD

**Discussion by hveiga** [apache/datafusion#10962] [Q&A] [answered]
Hi folks,

I am using Datafusion to partition some data stored in parquet files into a different set of parquet files. I would like those newly created files to contain the columns I am partitioning by, however currently the column gets removed as it becomes part of the file directory structure. Something like:

```
COPY (SELECT col1, col2, col3, col4 FROM my_external_table) TO '/output' PARTITIONED BY (col1) OPTIONS (format parquet);

...

/output/col1=val1/some_random_file_name.parquet
/output/col1=val2/some_random_file_name.parquet
/output/col1=val3/some_random_file_name.parquet
...

```

Is there a way in SQL to keep `col1` in the output parquet files? If not, would it make sense to add this as part of a new option when calling `COPY`? 

I have been looking in the documentation and open/closed issues but I could not find a way to do this, but if there is some information about it, a link would be greatly appreciated.

Thanks!

**Comment by devinjdangelo** [comment]
Your description of the behavior is correct. Datafusion will remove the partition columns from the data before writing out any files. There is not currently a config to modify this behavior, but it would be relatively straightforward to add. We would just need to make this line conditional on a config: https://github.com/apache/datafusion/blob/a8847e1a825f8e588362430d99363672d8c3e7db/datafusion/core/src/datasource/file_format/write/demux.rs#L303

**Reply by hveiga** [reply]
Understood. Thanks for the quick response. Is it ok to open an issue with this request? Or how should I proceed from here? I am happy to give it a try, but I must be honest I am not too familiar with the codebase.

**Reply by devinjdangelo** [reply, marked as answer]
> Understood. Thanks for the quick response. Is it ok to open an issue with this request? Or how should I proceed from here? I am happy to give it a try, but I must be honest I am not too familiar with the codebase.

Feel free to open a feature request issue! You are also free to take a stab at implementing if you are interested. I may be able to take a look at it myself as well.

**Reply by hveiga** [reply]
Created https://github.com/apache/datafusion/issues/10971 for this feature request. Thanks again!

**Reply by hveiga** [reply]
@devinjdangelo I would appreciate if you could take a look to https://github.com/apache/datafusion/issues/10971#issuecomment-2179646415 and give some guidance. Thanks!
